lmdeploy/serve/async_engine.py

```diff
@@ -648,25 +649,37 @@ def parse_tool_response(self, text, tools, **kwargs):
             name, parameters = action['name'], json.dumps(action.get(
                 'parameters', action.get('arguments', {})),
                                                           ensure_ascii=False)
+            call_info_list = [(name, parameters)]
         elif '<function=' in text:  # llama3.1
             action, _ = text.split('</function>')
             parameters = action[action.find('{'):]
             name = action.split('<function=')[1].split('>{')[0]
-        elif '<tool_call>' in text:  # qwen2.5
-            action = text.split('</tool_call>')[0]
-            action = action.split('<tool_call>')
-            if len(action) == 1:
-                text, action = '', action[0]
+            call_info_list = [(name, parameters)]
+        elif '<tool_call>' in text and '</tool_call>' in text:  # qwen2.5
+            # get tool_call in text
+            pattern = r'<tool_call>(.*?)</tool_call>'
+            match_result_list = re.findall(pattern, text, re.DOTALL)
+            call_info_list = []
+            for match_result in match_result_list:
+                action = json.loads(match_result)
+                call_info_list.append((action['name'],
+                                       json.dumps(action['arguments'],
+                                                  ensure_ascii=False)))
+            # get text outside of tags
+            if not text.startswith('<tool_call>'):
+                text = text[:text.find('<tool_call>')]
+            elif not text.endswith('</tool_call>'):
+                text = text[text.find('</tool_call>') + len('</tool_call>'):]
```

Comment:
Does it extract real text when there are multiple tool_calls in the response?

Reply:
sorry, it should be rfind